File: hydra-api/hydra/api/v1/models/ai.py
"""AI/LLM models for providers, configurations, and dynamic model fetching.

Terminology:
- LLMProvider: One of the 4 supported provider types (Anthropic, OpenAI, Ollama, OpenRouter)
- LLMModel: A model available from a provider (e.g., claude-3-5-sonnet, gpt-4o)
- LLMConfig: User's saved configuration (provider + model + API key) for a chat session
"""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class LLMProviderType(StrEnum):
    """Supported LLM provider types."""

    ANTHROPIC = "anthropic"
    OPENAI = "openai"
    OLLAMA = "ollama"
    OPENROUTER = "openrouter"


# =============================================================================
# LLM Configuration Models (User's saved provider+model configurations)
# =============================================================================


class LLMConfigCreate(BaseModel):
    """Request to create an LLM configuration."""

    name: str = Field(min_length=1, max_length=128, description="Display name for the configuration")
    type: LLMProviderType = Field(description="Provider type")
    api_key: str | None = Field(
        default=None,
        alias="apiKey",
        min_length=1,
        description="API key (not required for Ollama)",
    )
    base_url: str | None = Field(
        default=None,
        alias="baseUrl",
        description="Custom base URL (required for Ollama, optional for others)",
    )
    model: str = Field(
        min_length=1,
        max_length=128,
        description="Model to use (e.g., claude-3-5-sonnet-20241022)",
    )
    is_default: bool = Field(
        default=False,
        alias="isDefault",
        description="Set as the default configuration",
    )

    model_config = {"populate_by_name": True}


class LLMConfigUpdate(BaseModel):
    """Request to update an LLM configuration."""

    name: str | None = Field(default=None, max_length=128)
    api_key: str | None = Field(default=None, alias="apiKey", min_length=1)
    base_url: str | None = Field(default=None, alias="baseUrl")
    model: str | None = Field(default=None, max_length=128)
    is_default: bool | None = Field(default=None, alias="isDefault")

    model_config = {"populate_by_name": True}


class LLMConfigResponse(BaseModel):
    """LLM configuration response (API key masked)."""

    config_id: str = Field(alias="configId")
    name: str
    type: LLMProviderType
    api_key_last4: str | None = Field(alias="apiKeyLast4", description="Last 4 characters of API key")
    api_key_set: bool = Field(alias="apiKeySet", description="Whether an API key is configured")
    base_url: str | None = Field(alias="baseUrl")
    model: str
    is_default: bool = Field(alias="isDefault")
    is_valid: bool | None = Field(alias="isValid", description="Last validation result")
    last_validated_at: datetime | None = Field(alias="lastValidatedAt")
    created_by: str = Field(alias="createdBy")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}


class LLMConfigListResponse(BaseModel):
    """List of LLM configurations response."""

    configs: list[LLMConfigResponse]
    total: int


class LLMConfigValidateResponse(BaseModel):
    """Response for configuration validation."""

    config_id: str = Field(alias="configId")
    is_valid: bool = Field(alias="isValid")
    message: str
    validated_at: datetime = Field(alias="validatedAt")
    models: list[str] | None = Field(default=None, description="Available models if validation succeeded")

    model_config = {"populate_by_name": True}


# =============================================================================
# LLM Model Models (Models available from providers)
# =============================================================================


class LLMModel(BaseModel):
    """A model available from an LLM provider."""

    id: str = Field(description="Model identifier (e.g., claude-3-5-sonnet-20241022)")
    name: str = Field(description="Human-readable model name")
    context_window: int = Field(alias="contextWindow", description="Maximum context window in tokens")
    supports_tools: bool = Field(alias="supportsTools", description="Whether model supports tool/function calling")
    supports_vision: bool = Field(default=False, alias="supportsVision", description="Whether model supports image inputs")
    supports_reasoning: bool = Field(default=False, alias="supportsReasoning", description="Whether model supports extended thinking/reasoning")
    cost_per_1k_input: float | None = Field(default=None, alias="costPer1kInput", description="Cost per 1K input tokens in USD")
    cost_per_1k_output: float | None = Field(default=None, alias="costPer1kOutput", description="Cost per 1K output tokens in USD")

    model_config = {"populate_by_name": True}


class LLMModelsResponse(BaseModel):
    """Response containing available models from a provider."""

    models: list[LLMModel]
    provider: LLMProviderType
    fetched_at: datetime = Field(alias="fetchedAt")
    cached: bool = Field(default=False, description="Whether response was served from cache")

    model_config = {"populate_by_name": True}


# =============================================================================
# Global API Key Models
# =============================================================================


class GlobalKeyScope(StrEnum):
    """Scopes for global API key usage."""

    CHAT = "chat"  # Use for chat conversations
    META = "meta"  # Use for listing models, capabilities, costs
    TITLE_GEN = "title_gen"  # Use for auto-generating session titles


class GlobalAPIKeyCreate(BaseModel):
    """Request to create or update a global API key."""

    api_key: str = Field(alias="apiKey", min_length=1, description="API key for the provider")
    scopes: list[GlobalKeyScope] = Field(
        default=[GlobalKeyScope.CHAT, GlobalKeyScope.META, GlobalKeyScope.TITLE_GEN],
        description="Scopes this key should be used for",
    )

    model_config = {"populate_by_name": True}


class GlobalAPIKeyResponse(BaseModel):
    """Global API key response (key masked)."""

    provider_type: LLMProviderType = Field(alias="providerType")
    api_key_last4: str = Field(alias="apiKeyLast4", description="Last 4 characters of API key")
    api_key_set: bool = Field(alias="apiKeySet", description="Whether an API key is configured")
    scopes: list[GlobalKeyScope]
    is_valid: bool | None = Field(alias="isValid", description="Last validation result")
    last_validated_at: datetime | None = Field(alias="lastValidatedAt")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}


class GlobalAPIKeysResponse(BaseModel):
    """Response containing all global API keys for a user."""

    keys: list[GlobalAPIKeyResponse]


class GlobalKeyValidateResponse(BaseModel):
    """Response for global key validation."""

    provider_type: LLMProviderType = Field(alias="providerType")
    is_valid: bool = Field(alias="isValid")
    message: str
    validated_at: datetime = Field(alias="validatedAt")

    model_config = {"populate_by_name": True}

File: hydra-web/src/api/ai.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '@/lib/api-client';
import { queryKeys } from '@/lib/query-client';

export type LLMProviderType = 'anthropic' | 'openai' | 'ollama' | 'openrouter';

export interface LLMProviderResponse {
  configId: string;
  name: string;
  type: LLMProviderType;
  apiKeyLast4?: string | null;
  apiKeySet: boolean;
  baseUrl?: string | null;
  model: string;
  isDefault: boolean;
  isValid?: boolean | null;
  lastValidatedAt?: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface LLMProviderListResponse {
  configs: LLMProviderResponse[];
  total: number;
}

export interface LLMProviderCreate {
  name: string;
  type: LLMProviderType;
  model: string;
  apiKey?: string;
  baseUrl?: string;
  isDefault?: boolean;
}

export interface LLMProviderUpdate {
  name?: string;
  model?: string;
  apiKey?: string;
  baseUrl?: string;
  isDefault?: boolean;
}

export interface LLMProviderValidateResponse {
  configId: string;
  isValid: boolean;
  message: string;
  validatedAt: string;
  models?: string[] | null;
}

export function useLLMProviders() {
  return useQuery({
    queryKey: queryKeys.ai.models(),
    queryFn: async () => {
      const response = await apiClient.get<LLMProviderListResponse>('/ai/configs');
      return response.data;
    },
  });
}

export function useLLMProvider(providerId: string) {
  return useQuery({
    queryKey: queryKeys.ai.model(providerId),
    queryFn: async () => {
      const response = await apiClient.get<LLMProviderResponse>(`/ai/configs/${providerId}`);
      return response.data;
    },
    enabled: !!providerId,
  });
}

export function useCreateLLMProvider() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (data: LLMProviderCreate) => {
      const response = await apiClient.post<LLMProviderResponse>('/ai/configs', data);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.models() });
    },
  });
}

export function useUpdateLLMProvider() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async ({
      providerId,
      data,
    }: {
      providerId: string;
      data: LLMProviderUpdate;
    }) => {
      const response = await apiClient.put<LLMProviderResponse>(
        `/ai/configs/${providerId}`,
        data
      );
      return response.data;
    },
    onSuccess: (_data, { providerId }) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.model(providerId) });
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.models() });
    },
  });
}

export function useDeleteLLMProvider() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (providerId: string) => {
      const response = await apiClient.delete(`/ai/configs/${providerId}`);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.models() });
    },
  });
}

export function useValidateLLMProvider() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (providerId: string) => {
      const response = await apiClient.post<LLMProviderValidateResponse>(
        `/ai/configs/${providerId}/validate`
      );
      return response.data;
    },
    onSuccess: (_data, providerId) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.model(providerId) });
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.models() });
    },
  });
}

// =============================================================================
// Provider Models (Dynamic model fetching from provider APIs)
// =============================================================================

export interface LLMModel {
  id: string;
  name: string;
  contextWindow: number;
  supportsTools: boolean;
  supportsVision: boolean;
  supportsReasoning: boolean;
  costPer1kInput?: number | null;
  costPer1kOutput?: number | null;
}

export interface LLMModelsResponse {
  models: LLMModel[];
  provider: LLMProviderType;
  fetchedAt: string;
  cached: boolean;
}

export function useProviderModels(
  providerType: LLMProviderType,
  options?: { configId?: string; enabled?: boolean }
) {
  return useQuery({
    queryKey: queryKeys.ai.providerModels(providerType),
    queryFn: async () => {
      const params: Record<string, string> = {};
      if (options?.configId) {
        params.configId = options.configId;
      }
      const response = await apiClient.get<LLMModelsResponse>(
        `/ai/providers/${providerType}/models`,
        { params }
      );
      return response.data;
    },
    staleTime: 5 * 60 * 1000, // 5 minutes
    enabled: options?.enabled !== false,
  });
}

// =============================================================================
// Global API Keys (User's global API keys per provider with scopes)
// =============================================================================

export type GlobalKeyScope = 'chat' | 'meta' | 'title_gen';

export interface GlobalAPIKey {
  providerType: LLMProviderType;
  apiKeyLast4: string;
  apiKeySet: boolean;
  scopes: GlobalKeyScope[];
  isValid?: boolean | null;
  lastValidatedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface GlobalAPIKeysResponse {
  keys: GlobalAPIKey[];
}

export interface GlobalAPIKeyCreate {
  apiKey: string;
  scopes?: GlobalKeyScope[];
}

export interface GlobalKeyValidateResponse {
  providerType: LLMProviderType;
  isValid: boolean;
  message: string;
  validatedAt: string;
}

export function useGlobalKeys() {
  return useQuery({
    queryKey: queryKeys.ai.globalKeys(),
    queryFn: async () => {
      const response = await apiClient.get<GlobalAPIKeysResponse>('/ai/global-keys');
      return response.data;
    },
  });
}

export function useGlobalKey(providerType: LLMProviderType) {
  return useQuery({
    queryKey: queryKeys.ai.globalKey(providerType),
    queryFn: async () => {
      const response = await apiClient.get<GlobalAPIKey>(`/ai/global-keys/${providerType}`);
      return response.data;
    },
    enabled: !!providerType,
  });
}

export function useSetGlobalKey() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async ({
      providerType,
      data,
    }: {
      providerType: LLMProviderType;
      data: GlobalAPIKeyCreate;
    }) => {
      const response = await apiClient.put<GlobalAPIKey>(
        `/ai/global-keys/${providerType}`,
        data
      );
      return response.data;
    },
    onSuccess: (_data, { providerType }) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKey(providerType) });
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKeys() });
    },
  });
}

export function useDeleteGlobalKey() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (providerType: LLMProviderType) => {
      const response = await apiClient.delete(`/ai/global-keys/${providerType}`);
      return response.data;
    },
    onSuccess: (_data, providerType) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKey(providerType) });
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKeys() });
    },
  });
}

export function useValidateGlobalKey() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (providerType: LLMProviderType) => {
      const response = await apiClient.post<GlobalKeyValidateResponse>(
        `/ai/global-keys/${providerType}/validate`
      );
      return response.data;
    },
    onSuccess: (_data, providerType) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKey(providerType) });
      queryClient.invalidateQueries({ queryKey: queryKeys.ai.globalKeys() });
    },
  });
}
